Speed up CI "Initial JDK 11 Build" & fix some dependency gaps

famod opened the pull request:
- build in parallel
- fix "dependency gaps" via new update-extension-dependencies.sh support script
- don't execute any maven invoker tests

Will effectively use 2 threads which should speed things up by ~20%.

Plus: max maven repo compression level (also uses parallel compression)

gsmet commented on .github/workflows/ci-actions.yml:
```diff
@@ -83,7 +83,7 @@ jobs:
       - name: Build
         run: |
-          mvn -e -B -DskipTests -DskipITs -Dno-format clean install
+          mvn -T1C -e -B -DskipTests -DskipITs -Dinvoker.skip -Dno-format clean install
```
Ah nice, I wanted to experiment with that.

famod replied:
Too bad GH only gives us two cores so it's still rather slow (> 10 mins).

famod commented:
I rebased and aded another small improvement (maximum compression of the maven repo, in parallel).

famod commented:
Maven tests are failing with:
```
MojoExecutionException: Extensions referenced in io.quarkus:quarkus-bom:pom:999-SNAPSHOT and included in io.quarkus:quarkus-bom-quarkus-platform-descriptor:999-SNAPSHOT do not match expectations. 
```
With a long list of extensions missing from the descriptor:
```
Error:  Extensions from io.quarkus:quarkus-bom:pom:999-SNAPSHOT not present in io.quarkus:quarkus-bom-quarkus-platform-descriptor:999-SNAPSHOT
Error:  - io.quarkus:quarkus-spring-data-jpa:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-narayana-jta:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-reactive-db2-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-iam:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-scheduler:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-orm-panache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-orm-rest-data-panache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-reactive:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mongodb-panache-kotlin:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-quartz:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-sqs:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-reactive-messaging-mqtt:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-h2:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-picocli:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-rest-client-jaxb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-artemis-jms:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-mutiny-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-sns:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-minikube:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-grpc:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mutiny-reactive-streams-operators:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-postgresql:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elytron-security-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-derby:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-redis-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-jaxb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-openapi:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-dynamodb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-reactive-datasource:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-search-orm-elasticsearch:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-ses:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-reactive-panache-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-spring-cloud-config-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-validator:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-qute:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elytron-security-oauth2:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-orm:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-lambda-xray:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-grpc-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-container-image-openshift:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-datasource:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-reactive-messaging-kafka:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-agroal:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-micrometer:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-oidc:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mongodb-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-reactive-mysql-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-qute:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-rest-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-funqy-server-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-funqy-amazon-lambda:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-reactive-messaging-amqp:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-vertx-web:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mutiny:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-mssql:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-metrics:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jaxb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-google-cloud-functions-http:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-tika:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-s3:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jaeger:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-spring-scheduled:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-opentracing:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-undertow:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-health:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-jwt:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-funqy-http:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-rest-client-jsonb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-orm-panache-kotlin:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-vertx:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-jackson:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mongodb-rest-data-panache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-mutiny:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jsonb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-keycloak-authorization:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-azure-functions-http:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-context-propagation:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-openshift:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-security:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-mariadb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-artemis-core:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-webjars-locator:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-reactive-pg-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mongodb-panache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-neo4j:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-vertx-core:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-liquibase:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-cache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-fault-tolerance:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-logging-json:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elytron-security-ldap:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-vault:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jackson:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-jsonb:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elasticsearch-rest-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-undertow-websockets:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-consul-config:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-reactive-messaging:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-vertx-http:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-multipart:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elytron-security-properties-file:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-spring-security:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-graphql:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-reactive-panache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-keycloak-admin-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-funqy-knative-events:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-spring-web:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-lambda-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elasticsearch-rest-high-level-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-orm-panache-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-container-image-s2i:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mailer:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-mysql:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-rest-client-jackson:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-spring-cache:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-jdbc-db2:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-flyway:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-google-cloud-functions:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elytron-security-jdbc:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-kafka-streams:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-swagger-ui:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-lambda:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-smallrye-reactive-streams-operators:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-lambda-http:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-server-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-funqy-google-cloud-functions:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-kms:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-kubernetes:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-kubernetes-config:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-amazon-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-kubernetes-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-hibernate-envers:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-elytron-security:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-resteasy-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-infinispan-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-rest-client-mutiny:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-kafka-client:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-vertx-graphql:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-mongodb-panache-common:jar:999-SNAPSHOT
Error:  - io.quarkus:quarkus-security-jpa:jar:999-SNAPSHOT
```
I can't seem to reproduce this locally. But I am wondering whether `quarkus-bom-quarkus-platform-descriptor` needs a symbolic dependency to `quarkus-bom` (Maven does not consider the BOM import in `build-parent` as an actual dependency).

famod commented:
Well, I guess this snippet from `quarkus/pom.xml` explains it:
```xml
        <!-- Extensions -->
        <module>extensions</module>

        <!-- Devtools (Maven, cli, gradle) -->
        <!-- MUST BE BUILT AFTER EXTENSIONS BUT BEFORE IT -->
        <!-- the tools are generating the JSON descriptor for the BOM -->
        <!-- relying on the fact that the extension descriptor, generated as part of an extension build, is available -->
        <module>devtools</module>

        <!-- Integration Tests -->
        <module>integration-tests</module>
```
This is not trivial to solve. And it also affects local usage of `-T`...

gsmet commented:
`-T` works fine locally, I use it daily. I wonder if the issue is the change you did to `quarkus-cli`.

famod commented:
> `-T` works fine locally, I use it daily.

But you don't wipe your local maven repo, do you? Neither do I. But CI starts with an empty one AFAICT.

> I wonder if the issue is the change you did to `quarkus-cli`.

I thought so too in the beginning but I really think is a build order problem:

```
2020-11-11T09:48:38.1809088Z [INFO] ---------< io.quarkus:quarkus-bom-quarkus-platform-descriptor >---------
2020-11-11T09:48:38.1815310Z [INFO] Building Quarkus - BOM - Descriptor JSON 999-SNAPSHOT          [262/753]
[...]
2020-11-11T09:48:47.6375456Z [INFO] ------------------< io.quarkus:quarkus-logging-json >-------------------
2020-11-11T09:48:47.6376401Z [INFO] Building Quarkus - Logging - JSON - Runtime 999-SNAPSHOT       [263/753]
```
Which means that `quarkus-bom-quarkus-platform-descriptor` is built too early (because it does not depend on any/all extensions).

aloubyansky commented:
Right, concurrent builds won't work with an empty local repo, atm.

famod commented:
Input from @aloubyansky in PR #13243: https://github.com/quarkusio/quarkus/pull/13243#issuecomment-725950502
> The descriptor in fact does depend on the presence of the extensions it includes. So we could introduce dependencies at least on the POMs of all the extensions.

If we do this (which is very verbose, but the standard Maven-way) I wonder how we can make sure all extensions are listed as extensions? It should be possible via an enforcer rule.

Other solutions (with severe downsides):
- Split up the build job, but this is hard to enforce locally (and down goes developer joy).
- Add another Maven extension that auto-adds those dependencies. Major problem: Other extension like GIB won't see those dependencies (Maven limitation).
- Extend GIB. Slighty out of scope but somehow doable. But then again we also have non-GIB builds...

famod commented:
> Right, concurrent builds won't work with an empty local repo, atm.

Not only this but with a populated Maven repo you might run into inconsistencies in the platform descriptor because the generation might operate on outdated extension artifacts (the ones in the local repo and not the ones from the current reactor).

famod commented:
So, if we go the standard Maven-way:
- symbolic (type pom, scope test, optional, exclusions) or "real" dependencies? Technically we'd only need symbolic ones.
- dependencies to *-runtime or *-deployment? AFAICS only the descriptor in the runtime artifacts is relevant, no?
- do we want to add an enforcer rule?

aloubyansky commented:
Tbh, the ideal way would be to split the repo into the actual core and "core extensions".

aloubyansky commented:
Going forward I think we'll want to have the actual quarkus core bom that does not include any extension too.

aloubyansky commented:
Although, there would still be this issue with the descriptor dependencies.

aloubyansky commented:
We need the runtime extension artifacts for the descriptor.

famod commented:
> We need the runtime extension artifacts for the descriptor.

Ok, symbolic or real ones? I think I'd go for symbolic ones...

As for the enforcer-rule: Since there are tests already that are verifying the consistency, I'd say the such a rule is optional.

aloubyansky commented:
> > We need the runtime extension artifacts for the descriptor.
> 
> Ok, symbolic or real ones? I think I'd go for symbolic ones...

I don't have a strong preference. As long as it works.

> As for the enforcer-rule: Since there are tests already that are verifying the consistency, I'd say the such a rule is optional (or better to say not absolutely mandatory).

Right, there already is test that makes sure that all the extensions found in the BOM should also be in the descriptor and vice versa.

famod commented:
> Right, there already is test that makes sure that all the extensions found in the BOM should also be in the descriptor and vice versa.

Yeah. The only downside is that the root cause won't be crystal clear right away. I might try implementing a rule in a separate PR.

I'll update this PR with all the symbolic runtime dependencies.

gsmet commented:
Isn't this going to be a nightmare to maintain?

gsmet commented:
Frankly, it has already become very cumbersome to add a dependency to an extension, I wouldn't want things to become even more complicated.

If in the end, working in the tree becomes a pain, I'm not sure all this will be an improvement for our daily work.

famod commented:
> Isn't this going to be a nightmare to maintain?

Depends. It is certainly rather verbose. But you only have to change this one single pom.xml in case a new extension is added to this repo or an extension is removed. How often does that happen?
The only critical case is the addition of a new extension which will not be detected right away but will lead to an error in multiple subsequent jobs.

gsmet commented:
Isn't there a way to add these dependencies automatically via some Maven plugin?

And tbh I have the exact same question for the `-deployment` dependencies we add everywhere and make our poms far less readable (and error prone).

The fact that it doesn't happen often means everyone will forget about it and there's absolutely no value doing it manually.

maxandersen commented:
> Going forward I think we'll want to have the actual quarkus core bom that does not include any extension too.

that would make so much more sense to me - when can we have that ? :)

maxandersen commented:
+1 on looking into enforcer rule option if that is enough.

@gsmet we already discussed the value of having these extra metadata in https://github.com/quarkusio/quarkus/pull/11331 and agreed (as far as I remember) that since this can be automatically validated it is okey to live with extra verbosity since it enables things like telling mvn to build your current module and what it depends on instead of the whole quarkus project + of course the whole incremental build speed up in CI.

famod commented:
> +1 on looking into enforcer rule option if that is enough.

I have already been thinking about this a bit and in the end we would duplicate a good part of the logic of https://github.com/quarkusio/quarkus/blob/master/devtools/platform-descriptor-json-plugin/src/main/java/io/quarkus/maven/ValidateExtensionsJsonMojo.java
I wondered whether we could just call that mojo right after `bom-descriptor-json` but then again in a later CI job there are tests that ensure the mojo works as expected. 🤔 
A bit of a chickena nd egg situation...

famod commented:
I've added the dependencies to see how it goes.

FTR, quick'n'dirty:
```
for i in `grep -Po '(?<="artifact-id": ")[^"]+' target/quarkus-bom-quarkus-platform-descriptor-999-SNAPSHOT-999-SNAPSHOT.json`; do sed "s/XXX/$i/" template.xml; done > all.xml
```
`template.xml`:
```xml
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>XXX</artifactId>
            <version>${project.version}</version>
            <type>pom</type>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>*</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

famod commented:
CI is green.

@gsmet @aloubyansky @gastaldi how do we want to move on with this? Please see also https://github.com/quarkusio/quarkus/pull/13218#issuecomment-726211189.

gastaldi commented:
@famod you might have probably tested that, but will this impact when we start using GIB?

gastaldi reviewed (approved):
LGTM; let's see how this behaves

famod commented:
> @famod you might have probably tested that, but will this impact when we start using GIB?

It will _also_ impact GIB in that if just an extension has changes the descriptor will also be built because it is detected as a downstream project.
But it already has an impact now without GIB.

gsmet reviewed (changes requested):
While stabilizing the parallel build is good (thanks!), I think we need a script to maintain this. The enforcer rule is OK but I think having a simple script generating everything and replacing the **sorted** list in a `<-- Extension dependencies -->...<-- /Extension dependencies -->` would be beneficial.

Especially since we will probably need to add it in several places (the JSON descriptor is one but the documentation is another).

The enforcer rule could simply say `Please run ./update-extension-dependencies.sh` and the script would fix it in every place where it's needed. I don't think it would be that hard (especially since @famod already did it with a script). For now, we can go with a simple bash script, it's good enough. If the OP cannot execute it, a committer can run it and update the PR.

@maxandersen As for agreeing about verbosity, frankly, if it's becoming worse and worse and we add more and more of this noise, it's a totally different matter. I was already reluctant on the first batch, I haven't signed for adding more without having the tools to not make it a nightmare to maintain.

maxandersen commented:
Fully agree @gsmet to what you are saying. @famod already suggested to automate it as much as possible so I think we are all in agreement that we want that.

famod commented:
I get the point of providing a script to aid with updating the dependency list for `bom-descriptor-json`. I was actually thinking about adding the command as a comment to the pom.xml but sure, why not provide ready-to-use script. :+1:

Such a script should ideally update the respective pom.xml _directly_, right? (Not create a "foo.xml" that needs to be copied _manually_ into the right element.)

What leaves a question or two is the enforcer rule. How are we going to solve [the chicken and egg problem](https://github.com/quarkusio/quarkus/pull/13218#issuecomment-726211189)?
Or is this
> For now, we can go with a simple bash script, it's good enough.

implying that the enforcer rule is something for a later PR? Just trying to not misunderstand things here. 😉

aloubyansky commented:
The good thing about validating the json in the IT is that by that time we can (should) be sure that all the extensions have been built, which is the prerequisite for validation.

famod commented:
I've pushed a first version of `update-extension-dependencies.sh`. I converted this PR to a draft until we have finished polishing it.

I'll add a few comments to the file.

WDYT?

famod commented on update-extension-dependencies.sh:
```diff
@@ -0,0 +1,45 @@
+#mvn -D quickly -T0.8C -f independent-projects/arc
+#mvn -D quickly -T0.8C -f bom/application
+#mvn -D quickly -T0.8C -f core
+#mvn -D quickly -T0.8C -f extensions
```
This and the previous commands would make sure that everything relevant that goes into the descriptor is up to date.
On the other hand we could instead just _expect_ everything to be up to date and remove those commands.

PS: Building everything _except_ `integration-tests` is not that easy, that's why multiple commands...

famod replied:
Update: I removed those lines.

famod commented on update-extension-dependencies.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
Placement in root ok? Or somewhere else?

gastaldi replied:
That is fine 👍🏻

famod commented on update-extension-dependencies.sh:
```diff
@@ -0,0 +1,45 @@
+#mvn -D quickly -T0.8C -f extensions
+mvn clean package -f devtools/bom-descriptor-json
+
+DEP_TEMPLATE='        <dependency>
```
General idea is to avoid additional (temporary) files, so everything is stored in variables.

famod commented on update-extension-dependencies.sh:
```diff
@@ -0,0 +1,45 @@
+  | xargs -i sh -c "echo \"${DEP_TEMPLATE}\" | sed 's/XXX/{}/'" \
+  | sed ':a;N;$!ba;s/\n/\\\n/g'`    # replace newlines with \n so that sed accepts ${DEPS} as input
+
+MARK_START='<!-- START update-extension-dependencies.sh -->'
```
I think it is a good idea to have in the pom.xml a reference to the script.

famod commented on update-extension-dependencies.sh:
```diff
@@ -0,0 +1,45 @@
+echo ''
+echo 'Updating bom-descriptor-json/pom.xml...'
+echo ''
+sed -i "/${MARK_START}/,/${MARK_END}/c\        ${MARK_START}\n${DEPS}\n        ${MARK_END}" devtools/bom-descriptor-json/pom.xml
```
More pom.xml could be added here although I'm asking myself is whether `docs` should really receive the very same deps as `bom-descriptor-json`?

famod replied:
I've just pushed an update that includes `docs`. 🤷‍♂️

famod commented:
@gsmet WDYT? I'd like to finish this PR. Thanks!

famod commented:
I finalized this as I saw fit. I also added a small section to `CONTRIBUTING.md`.

@gsmet please re-review. Thanks!

famod commented:
Just rebased. @gsmet: friendly reminder. 🙂 Thanks!

gsmet commented:
Yeah, I have that one in mind. Just unsure when I will have some quality time to dedicate to this. I'll ping you as soon as I do.

famod commented:
Just a thought: The new script could be run by CI and in case modified pom.xml files are found after that, then someone forgot to run the script locally (or forgot to commit the updated pom.xml files) and we could fail the job.

famod commented:
@gsmet Please let me know if/when you need a rebase pushed.
I'm rebasing this every now and then and in case the script adds new dependencies, I will push right away, but the last few days were silent in that regard.

famod commented:
I removed my change from quarkus-cli because this has been taken care of by #14042.

famod commented:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  07:32 min (Wall Clock)
[INFO] Finished at: 2021-01-04T17:54:09Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:java (default) on project quarkus-documentation: An exception occured while executing the Java class. Failed to resolve artifacts: Could not find artifact io.quarkus:quarkus-spring-data-rest-deployment:jar:999-SNAPSHOT -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:java (default) on project quarkus-documentation: An exception occured while executing the Java class. Failed to resolve artifacts
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
Caused by: org.apache.maven.plugin.MojoExecutionException: An exception occured while executing the Java class. Failed to resolve artifacts
    at org.codehaus.mojo.exec.ExecJavaMojo.execute (ExecJavaMojo.java:311)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
Caused by: io.quarkus.bootstrap.resolver.maven.BootstrapMavenException: Failed to resolve artifacts
    at io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver.resolve (MavenArtifactResolver.java:183)
    at io.quarkus.docs.generation.AllConfigGenerator.main (AllConfigGenerator.java:110)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:834)
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact io.quarkus:quarkus-spring-data-rest-deployment:jar:999-SNAPSHOT
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:424)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifacts (DefaultRepositorySystem.java:270)
    at io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver.resolve (MavenArtifactResolver.java:181)
    at io.quarkus.docs.generation.AllConfigGenerator.main (AllConfigGenerator.java:110)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:834)
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact io.quarkus:quarkus-spring-data-rest-deployment:jar:999-SNAPSHOT
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:414)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifacts (DefaultRepositorySystem.java:270)
    at io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver.resolve (MavenArtifactResolver.java:181)
    at io.quarkus.docs.generation.AllConfigGenerator.main (AllConfigGenerator.java:110)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:834)
```
Seems I have to use `*-deployment` deps instead of the runtime ones, at least for `docs`. I'll try to adjust the script.

[ ] use deployment deps

famod commented:
@gsmet It would be good to finalize this one as people are stumbling upon those gaps every now and then, e.g. https://github.com/mvndaemon/mvnd/issues/264#issuecomment-756132459

n1hility commented:
Thanks for updating the contribution docs as part of this change.

famod commented:
JVM Tests - JDK 11 build failure is caused by a flaky test: #14392

famod commented:
@gsmet a new week, another friendly review reminder. 😉

famod committed:
Speed up CI "Initial JDK 11 Build" & fix some dependency gaps

- build in parallel
- fix "dependency gaps" via new update-extension-dependencies.sh support script
- don't execute any maven invoker tests

famod committed:
CI: max maven repo compression level

also uses parallel compression

gsmet reviewed (approved):
OK, let's merge this one and see how it goes, it looks sane to me! Sorry for the delay!
